src/output-filestore.c

```diff
@@ -0,0 +1,482 @@
+                SCLogNotice("failed to open file %s: %s", filename, strerror(errno));
+                return -1;
+            }
+        } else {
+            file_fd = ff->fd;
+        }
+    }
+
+    if (file_fd != -1) {
+        ssize_t r = write(file_fd, (const void *)data, (size_t)data_len);
+        if (r == -1) {
+            SCLogDebug("write failed: %s", strerror(errno));
+            if (ff->fd != -1) {
+                SC_ATOMIC_SUB(filestore_open_file_cnt, 1);
+            }
+            ff->fd = -1;
+        }
+        if (ff->fd == -1) {
+            close(file_fd);
+        }
+    }
+
+    if (flags & OUTPUT_FILEDATA_FLAG_CLOSE) {
+        if (ff->fd != -1) {
+            close(ff->fd);
+            ff->fd = -1;
+            SC_ATOMIC_SUB(filestore_open_file_cnt, 1);
+        }
+        OutputFilestoreFinalizeFiles(ctx, p, ff);
+    }
+
+    return 0;
+}
+
+static TmEcode OutputFilestoreLogThreadInit(ThreadVars *t, const void *initdata,
+        void **data)
+{
+    OutputFilestoreLogThread *aft = SCMalloc(sizeof(OutputFilestoreLogThread));
+    if (unlikely(aft == NULL))
+        return TM_ECODE_FAILED;
+    memset(aft, 0, sizeof(OutputFilestoreLogThread));
+
+    if (initdata == NULL) {
+        SCLogDebug("Error getting context for LogFileStore. \"initdata\" argument NULL");
+        SCFree(aft);
+        return TM_ECODE_FAILED;
+    }
+
+    OutputFilestoreCtx *ctx = ((OutputCtx *)initdata)->data;
+    aft->ctx = ctx;
+
+    aft->counter_max_hits =
+        StatsRegisterCounter("file_store.open_files_max_hit", t);
+
+    *data = (void *)aft;
+    return TM_ECODE_OK;
+}
+
+static TmEcode OutputFilestoreLogThreadDeinit(ThreadVars *t, void *data)
+{
+    OutputFilestoreLogThread *aft = (OutputFilestoreLogThread *)data;
+    if (aft == NULL) {
+        return TM_ECODE_OK;
+    }
+
+    /* clear memory */
+    memset(aft, 0, sizeof(OutputFilestoreLogThread));
+
+    SCFree(aft);
+    return TM_ECODE_OK;
+}
+
+static void OutputFilestoreLogExitPrintStats(ThreadVars *tv, void *data)
+{
+    OutputFilestoreLogThread *aft = (OutputFilestoreLogThread *)data;
+    if (aft == NULL) {
+        return;
+    }
+
+    SCLogInfo("(%s) Files extracted %" PRIu32 "", tv->name, aft->file_cnt);
+}
+
+static void OutputFilestoreLogDeInitCtx(OutputCtx *output_ctx)
+{
+    OutputFilestoreCtx *ctx = (OutputFilestoreCtx *)output_ctx->data;
+    SCFree(ctx);
+    SCFree(output_ctx);
+}
+
+static void GetLogDirectory(const ConfNode *conf, char *out, size_t out_size)
+{
+    const char *log_base_dir = ConfNodeLookupChildValue(conf, "dir");
+    if (log_base_dir == NULL) {
+        SCLogNotice("Using default log directory %s", default_log_dir);
+        log_base_dir = default_log_dir;
+    }
+    if (PathIsAbsolute(log_base_dir)) {
+        strlcpy(out, log_base_dir, out_size);
+    } else {
+        const char *default_log_prefix = ConfigGetLogDirectory();
+        snprintf(out, out_size, "%s/%s", default_log_prefix, log_base_dir);
```

Comment:
wonder if we need a string constructor func for paths, so that we can create clean windows paths as well

Reply:
though it seems the current path works on windows (mingw build)

Reply:
Apparently Windows has supported the posix path separator since Windows XP. News to me, but handy.

Reply:
But say you specify the default log dir as "c:\ids\logs", and we tack on "filestore" resulting in "c:\ids\logs/filestore". I wonder if thats a problem?